### PR TITLE
Add support for compressed HTTP requests (gzip, deflate)

### LINKED ISSUE #1135: Does not specify the right headers in Access-Control-Allow-Headers
## Expected Behavior
Trying to execute a `kuzzle.index.exists('foo')` in the **browser**, which should at least trigger a request to Kuzzle.

## Current Behavior
I'm getting a `Request header field x-kuzzle-volatile is not allowed by Access-Control-Allow-Headers in preflight response.` error in the console.

## Possible Solution
It would be great that Kuzzle specifies the right allowed headers when asked by the `connect()` call. Currently, the only allowed headers are `Content-Type, Access-Control-Allow-Headers, Authorization, X-Requested-With`.

## Steps to Reproduce
1. Load the 6.x Kuzzle SDK in a web page
2. call `connect()`
3. call any other method (my case was `kuzzle.index.exists('foo')`)

### PR DESCRIPTION
**Pull request by scottinet**
:warning: HTTP Proxy functional tests will fail until https://github.com/kuzzleio/kuzzle-proxy/pull/100 is merged (and docker images are updated)

## What does this PR do?

* Add support for incoming compressed HTTP requests (accepted algorithms: gzip, deflate). As per the RFC, the `content-encoding` headers contains the *list* of compression algorithms used, in the order used to compress the data (a request may be compressed with multiple layers of gzip and/or deflate)
* Add support for outgoing compressed HTTP requests. This is controlled by the requesting client, using the `accept-encoding` header. Accepted algorithms: gzip, deflate. Because deflate is often badly implemented by clients, gzip is preferred when given a choice 
* The HTTP headers sent by default by Kuzzle now contain an `accept-encoding` headers
* Two new configuration parameters have been added allowing administrators to control this feature:
  * `server.protocols.http.allowCompression` (boolean, default: `true`): used to enable/disable the feature for *incoming* requests
  * `server.protocols.http.maxEncodingLayers` (integer, default: `3`): limits the number of compression layers, to prevent cpu/memory bombs

### How should this be manually tested?

New HTTP-only functional tests have been added, automatically testing the feature.
If you want to check manually, use your favorite HTTP client and play with the `content-encoding` and `accept-encoding` headers

### Other changes

* Reading HTTP incoming messages is now properly done, using stream pipes, instead of emulating that feature with events
* Promoted the `entryPoint` property in the protocol implementations in the abstract Protocol class, DRYifying it

### Boyscout

* Add more extensive HTTP options documentation in the `.kuzzlerc.sample` file
* Add missing headers in the `Access-Control-Allow-Headers` list (fix #1135)

### CODE REVIEW

File: lib/api/core/entrypoints/embedded/protocols/http.js [hunk 1]
```diff
@@ -287,9 +277,143 @@ class HttpProtocol extends Protocol {
...
+      // this means that we need to invert the list to correctly
+      // decode the message
+      encodings.reverse();
+
+      if (encodings.length > this.maxEncodingLayers) {
+        throw new BadRequestError('Too many encodings');
+      }
+
+      for (const encoding of encodings) {
+        let pipe = null;
```
Comment (kuzzle) [inline lib/api/core/entrypoints/embedded/protocols/http.js]: MAJOR Remove this useless assignment to local variable "pipe" [rule](http://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1854)

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: lib/api/core/httpRouter/index.js [hunk 2]
```diff
@@ -52,11 +52,16 @@ class Router {
 
     this.defaultHeaders = {
       'content-type': 'application/json',
+      'Accept-Encoding': 'identity',
       'Access-Control-Allow-Origin': kuzzle.config.http.accessControlAllowOrigin || '*',
       'Access-Control-Allow-Methods': 'GET,POST,PUT,DELETE,OPTIONS,HEAD',
-      'Access-Control-Allow-Headers': 'Content-Type, Access-Control-Allow-Headers, Authorization, X-Requested-With'
+      'Access-Control-Allow-Headers': 'Content-Type, Access-Control-Allow-Headers, Authorization, X-Requested-With, Content-Encoding'
```
Comment (ballinette) [inline lib/api/core/httpRouter/index.js]: not directly related with the issue this PR resolves... but with @xbill82 we noticed that we need also to add `x-kuzzle-volatile` and  `Content-Length` to allow-headers. (see #1135 )

Comment (scottinet) [reply lib/api/core/httpRouter/index.js]: Of course!
Headers updated :+1:

**Pull request merged**